Week4/LESSON_PLAN.md

```diff
@@ -0,0 +1,275 @@
+    reporterId: ObjectId("e810507f191de86032c19729"),
+  }, {
+    timestamp: new Date('2000-06-06T12:34:00'),
+    page: "home",
+    stacktrace: "", // removed for clarity
+    reporterId: ObjectId("e810507f191de86032c19729"),
+  }
+];
+```
+
+#### Sort
+
+If we want to sort the find query on the timestamp to find the latest bugs we can run the following query:
+
+```js
+client.db("logging").collection("bugs").find().sort({ timestamp: -1 });
+```
+
+#### Limit
+
+The above query will give you back all of the bugs which is not great, as such we can use the limit command as follows:
+
+```js
+client
+  .db("logging")
+  .collection("bugs")
+  .find()
+  .sort({ timestamp: -1 })
+  .limit(10);
+```
+
+This will give only the last 10 bugs, which is more manageable. Note that you can put the `sort` and `limit` commands in any order!
+
+#### Aggregate
+
+Let's say we want to have a count of how many bugs appear per page. To do that we can do the following:
+
+```js
+client
+  .db("logging")
+  .collection("bugs")
+  .aggregate([
+    {
+      $group: {
+        _id: "$page",
+        count: { $count: {} },
+      },
+    },
+  ]);
+```
+
+This will give back an object with the page field in the `_id` field and the number of bugs that were logged on that page is in the `count` field!
+
+### Exercise
+
+Think of some other things to sort on or calculate and write the code to do that.
+
+### Essence
+
+MongoDB does a lot for you, the syntax is a little different than known, but the documentation is very detailed so make use of it!
+
+## 3. Pagination
+
+Using the same bugs collection, let's look at offset and cursor-based pagination using that collection.
+
+### Explanation
+
+Go to different online shops and look at their results pages to show pagination in action.
+
+### Example
+
+Given the bugs database in the previous section let's implement both types of pagination:
+
+#### Offset-based
+
+```js
+client
+  .db("logging")
+  .collection("bugs")
+  .find()
+  .sort({ timestamp: -1 })
+  .limit(10)
+  .skip(20);
+```
+
+This would skip 20 results and then show the next 10. So would be on page 3 if we show 10 results per page!
+
+#### Cursor-based
+
+```js
+const latestBugs = await client
+  .db("logging")
+  .collection("bugs")
+  .find({
+    timestamp: { $lt: next || new Date() },
+  })
+  .sort({ timestamp: -1 })
+  .limit(10);
+
+const cursorToGiveToUser = latestBugs[latestBugs.length - 1].timestamp;
+```
+
+Two important things here:
+
+- You need to always have the data sorted if you do cursor-based sorting as you are including the point you are at in the query.
+- You have to provide the user of your endpoint the information they need to send for the next query
+
+In the above code we do it with the timestamp, in other implementations an ID can be given.
+
+### Exercise
+
+Discuss the advantages and disadvantages of both approaches.
+
+### Essence
+
+Pagination needs to be done, important to note is that it only works with sorted data!
```

Comment:
The essence of pagination is that it may be useful but not necessarily with `skip-limit` always
[like explained here](https://www.mongodb.com/blog/post/paging-with-the-bucket-pattern--part-1)